### ISSUE TITLE
MIDI: noteOn message with velocity 0 is parsed as noteOff message

### ISSUE THREAD

**Issue by sensestage** [supercollider/supercollider#1483]
In AlsaMIDI a noteOn message with velocity 0 is parsed as noteOff message automatically.
It would be better to have this as an option configured from the language, as currently if I define a noteOn action, it will never fire for noteOn actions with velocity 0.

lines 179 and up of SC_AlsaMIDI.cpp:
```
			case SND_SEQ_EVENT_NOTEON:			// noteOn
				++g->sp; SetInt(g->sp, evt->data.note.channel);
				++g->sp; SetInt(g->sp, evt->data.note.note);
				++g->sp; SetInt(g->sp, evt->data.note.velocity);
				runInterpreter(g, evt->data.note.velocity ? s_midiNoteOnAction : s_midiNoteOffAction, 5);
				break;
```

**Comment by sensestage** [comment]
in CoreMIDI as well:

```
				case 0x90 : //noteOn
					++g->sp; SetInt(g->sp, pkt->data[i+1]); //val1
					++g->sp; SetInt(g->sp, pkt->data[i+2]); //val2
					runInterpreter(g, pkt->data[i+2] ? s_midiNoteOnAction : s_midiNoteOffAction, 5);
					i += 3;
					break;
```

**Comment by sensestage** [comment]
and PortMIDI:

```
					case 0x90 : //noteOn 
						++g->sp; SetInt(g->sp, data1);
						++g->sp; SetInt(g->sp, data2);
						runInterpreter(g, data2 ? s_midiNoteOnAction : s_midiNoteOffAction, 5);
						break;
```

**Comment by sensestage** [comment]
Alternatively, in order not to break existing code, it should fire both.

**Comment by telephon** [comment]
I found this comment, by AdmiralQuality:
http://www.kvraudio.com/forum/viewtopic.php?t=291892

"There's a thing in MIDI called "running status". Normally a MIDI message consists of a status byte (which always begins with the high order bit set to one: 1xxxxxxx), followed by 2 data bytes (which always have the high order bit set to zero: 0xxxxxxx).

The status byte is used to describe the type of message, there's one code for note-on, another for note-off, another for control change, etc...

Then the first data byte describes WHICH control or note has been changed. i.e. the CC# or the note #. Because we never use the high-bit, we get a range of 0 - 127.

And the second data byte describes the position of the control or the velocity of the note. Again we have a range of 0 - 127 (however for note velocities, zero is reserved to mean note off because...)

If a whole bunch of messages come through, all for the same status, MIDI has the ability to save a some bandwidth by leaving out the status byte. This is called "running status". And obvious use for it is when performing a control change, you can save 1/3 of the traffic on the MIDI cable by only sending the status byte at the beginning of the stream (note that this works only as long as the stream isn't interrupted by any other type of message status. As soon as another type of message comes through, the running status is broken).

And to take advantage of this for note data, a stream of note-ons and note-offs can be differentiated by the velocity value in the 2nd data byte, without needing to send a new status byte every time your performance switches from note-on to note-off.

So by the MIDI spec, you are required to treat a note-on with zero velocity exactly the same as you would a note-off."

**Comment by telephon** [comment]
But, reading on, I realise it is disputed. Maybe we need a setting for it.

**Comment by sensestage** [comment]
It is disputed - and I'd rather be able to parse it on my own, for my own usecase, rather than having to listen to both types of messages.

If MIDIIn just has a flag to: translateNoteOnZeroToOff with a default of true, then it can be handled by the user, if she wants to.

I can work on a patch to fix this.

**Comment by sensestage** [comment]
Also - a programming language interpreter should leave the interpretation of a message to the (SC) programmer, and not translate incoming data before it is accessible in the language.

**Comment by telephon** [comment]
>Also - a programming language interpreter should leave the interpretation of a message to the (SC) programmer, and not translate incoming data before it is accessible in the language.

Yes, exactly.

**Comment by jamshark70** [comment]
1. This is not a critical bug fix, so, not for 3.7.

2. We should weigh the risk of breaking people's code against the real pain that this issue is causing to users. In my 12 years in SuperCollider-land, nobody has complained about note-on messages with 0 velocity changing automatically into note-offs -- not even once. User pain just isn't there. This issue is talking about fixing something that nobody really thinks is broken. IMO it's a waste of time, while some 3.7 showstopper bugs are not being fixed quickly.

**Comment by telephon** [comment]
OK, so just for the record, the default should then be ````zeroNoteOnAsNoteOff = true```` (or the like).

**Comment by sensestage** [comment]
I just made a pull request.

**Comment by sensestage** [comment]
I really would like to get this into 3.7 - it is a feature that would make the Modality toolkit (which is getting close to a release) more transparent.

**Comment by LFSaw** [comment]
I opt for this. 
Just for the records, this is Modality: http://modality.bek.no

**Comment by jamshark70** [comment]
If there is a concrete use case (e.g. Modality), then I don't object. But you can understand why the issue as initially logged looked like pointless tinkering.

This is going to make VoicerMIDISocket a bit trickier, but, OK.

**Comment by sensestage** [comment]
No, the default behaviour stays as it was.

**Comment by jamshark70** [comment]
If Modality throws the switch to change the behavior, then my MIDI library needs to handle note-offs that happen to be sent as note-on with velocity = 0. In VoicerMIDISocket, the result of not handling this will be two-fold: notes won't release, and the 0-velocity messages will cause zombie nodes on the server.

So either it's trickier, or I'd have to declare that VoicerMIDISocket is incompatible with Modality.

It's not hard for me to fix, just ugly (less elegant).

**Comment by LFSaw** [comment]
Nice that you want to make your library compatible with our Modality endeavour, @jamshark70 !
Personally, I also prefer having a clear separation between noteOn and noteOff messages, so I am sure that we'll make sure that Modality will be compatible with both modes.

**Comment by timblechmann** [comment]
will this change break user-code?

**Comment by LFSaw** [comment]
@timblechmann no.

**Comment by scztt** [comment]
It sounds like the "vel:0 as note off" behavior is hardware specific (i.e. some devices follow that pattern, some don't). Shouldn't zeroNoteOnAsNoteOff be an instance property of MIDIIn, rather than a classvar? Since I might have one device that's sending be 0's at the end of every note, and another that's not?

**Comment by jamshark70** [comment]
MIDIIn is a low-level abstraction for the inputs coming from the OS's MIDI backend. There are no instances of it.

**Comment by LFSaw** [comment]
... which means that, if you set Meta_MIDIIn:zeroNoteOnAsNoteOff_(true), you may write a plugin to the higher-level abstractions (such as to MIDIIn instances) which does exactly that behaviour. With the current implementation, this is not possible.

**Comment by telephon** [comment]
In principle, it probably should be a property of ````MIDIMessageDispatcher````. If I understand correctly, the class method for MIDIIn is really a compromise for backward compatibility.

**Comment by sensestage** [comment]
Indeed. the class method is a compromise for backward compatibility.
And yes, MIDIIn only has class methods to handle the data coming in from the primitives.

I think Modality would only set the flag once it opens a device that uses noteOns.

But it might be cleaner to make a MIDIMessageDispatcher that does the translation, to have it configurable on a case-by-case basis. That dispatcher could then be passed in, if needed for co-existance of both features.

It gets kind of ugly - in order not to break user code, we need the flag at the MIDIIn level, since old code that uses MIDIIn directly may rely on the automatic translation.
So then we would have the flag/dispatcher at two places...

Or - since it is a major release, we should be really clear about it, and just provide the dispatcher if people want the old behaviour, and people using MIDIIn directly need to change their code.
(otherwise we cannot change the default till 3.8).

**Comment by telephon** [comment]
>Or - since it is a major release, we should be really clear about it, and just provide the dispatcher if people want the old behaviour, and people using MIDIIn directly need to change their code.
(otherwise we cannot change the default till 3.8).

Somehow, I tend to agree. But then we need to announce it LARGE enough so that it is probable that someone will notice it who doesn't even remember that old piece she made where this behavior was taken for granted.

**Comment by patrickdupuis** [comment]
This issue can be closed? @crucialfelix

**Comment by crucialfelix** [comment]
What was the consensus ? To implement this in other libraries and leave core alone ?  Somebody would need to take action or we should close it as stale.

**Comment by jamshark70** [comment]
While I'm aware of the reasons to change them, I have to admit, I'm not convinced, but I won't stand in the way, either.

A MIDIIn switch is, on second thought, not a good idea. Suppose Modality expects it to be set one way while my MIDI quark expects the other. It can be set only one way at one time, so somebody's code will be broken.

A MIDI message dispatcher may be a better approach.

But to be honest, when does anybody actually want a note-on with velocity 0? It sounds to me like it's just a matter of principle without a concrete application. Again, if others feel strongly that we just cannot live without note-on plus 0 velocity, I won't stand in the way, but I'm really rather skeptical of this. I have doubts about breaking stuff just because of an icky feeling about a decades-old MIDI convention.

My first vote: Leave it alone.
Secondary vote: Put a switch in the MIDI dispatcher.

**Comment by crucialfelix** [comment]
Its nearly a year old and it's been debated. Closing. Re-open if the passion to discuss and implement returns. thx

**Comment by muellmusik** [comment]
As a brief post-mortem: James is right that it would be trivial to implement a custom or alternate dispatcher for MIDIFunc to handle this. But actually, the following is already possible, and not too onerous:
```
~offFunc = {...};
m = MIDIFunc.noteOff(~offFunc);
n = MIDIFunc.noteOn(~offFunc, argTemplate: 0);
```
If you're not receiving noteOffs it's even simpler. Of course if you're just using the low level hook, and writing your own handlers that doesn't help you, but if you're doing that you're already mucking about in the guts anyway.

**Comment by LFSaw** [comment]
As mentioned earlier in this thread, a flag was introduced with f8c46e5 (see PR #1488 ) which can be set by the user. It is currently set such that the default behaviour is unchanged.

**Comment by crucialfelix** [comment]
Its very hard to tell at a glance from this thread. I didn't notice any mention of a fix. But I don't have half an hour to read each and every thread :) Thanks for the report, LF

**Comment by LFSaw** [comment]
no-one asked you to read the whole thing, I just remembered it so I though I'd mention :)
(the invitation for a beverage of you choice is still on...)

**Comment by crucialfelix** [comment]
Yes that was an honest thank you for clarifying !

**Comment by jamshark70** [comment]
Then I vote for reverting the pull request.

The point of having a user setting is so that a library can set it and assume correct behavior. Having only one global setting for the whole system means that, for example, if I set it the way I like, then Modality breaks, and if Modality sets it the way Marije likes, then my stuff breaks... unless we both implement alternate code branches depending on the setting... but the setting was supposed to avoid requiring the user to write alternate code branches!

I apologize that I wasn't paying attention to the PR but I highly emphatically disagree with this solution. IMO get rid of it before people start depending on it.

**Comment by jamshark70** [comment]
To clarify: I won't stand in the way of a mechanism to allow users to decide which behavior they want, provided that the mechanism does not needlessly complicate user code or create unresolvable conflicts. The current solution doesn't accomplish that.

**Comment by muellmusik** [comment]
> but the setting was supposed to avoid requiring the user to write alternate code branches!

This is a good point. We generally don't give enough consideration to how things will interact with library code. This is particularly true of global switches.

**Comment by muellmusik** [comment]
I think the library interaction question needs to be resolved, so I've reopened.

**Comment by crucialfelix** [comment]
Actually I would suggest opening a new issue. The previous issue was resolved with a pull request and was released in 3.7.  That's history - it's done.

Now you want to discuss reverting that and the actions to take to do that. That should be a new issue that references this one.

**Comment by LFSaw** [comment]
@sensestage should be part if this conversation...
